spec/lib/stellar/sep10_spec.rb

```diff
@@ -0,0 +1,604 @@
+      op.source_account = nil
+
+      expect { 
+        sep10.read_challenge_tx(challenge_xdr: envelope.to_xdr(:base64), server: server)
+      }.to raise_error(Stellar::InvalidSep10ChallengeError, /The transaction's operation should contain a source account/)
+      end
+      
+    it "throws an error if operation is not manage data"  do
+      envelope.tx.operations = [ 
+        Stellar::Operation.payment(
+          destination: Stellar::KeyPair.random, 
+          amount: [:native, 20],
+          source_account: Stellar::KeyPair.random
+        )
+      ]
+
+      expect { 
+        sep10.read_challenge_tx(challenge_xdr: envelope.to_xdr(:base64), server: server)
+      }.to raise_error(Stellar::InvalidSep10ChallengeError, /The transaction's operation should be manageData/)
+      end
+
+    it "throws an error if operation value is not a 64 bytes base64 string" do
+      transaction.operations[0].body.value.data_value = SecureRandom.random_bytes(64)
+      expect { 
+        sep10.read_challenge_tx(challenge_xdr: envelope.to_xdr(:base64), server: server)          
+      }.to raise_error(
+        Stellar::InvalidSep10ChallengeError,
+        /The transaction's operation value should be a 64 bytes base64 random string/
+      )
+    end
+
+    it "throws an error if transaction is not signed by the server" do
+      envelope.signatures = envelope.signatures.slice(1, 2)
+      
+      expect { 
+        sep10.read_challenge_tx(challenge_xdr: envelope.to_xdr(:base64), server: server)          
+      }.to raise_error(
+        Stellar::InvalidSep10ChallengeError,
+        /The transaction is not signed by the server/
+      )
+    end
+
+
+    it "throws an error if transaction does not contain valid timeBounds" do
+      envelope.tx.time_bounds = nil
+      challenge = envelope.tx.to_envelope(server, user).to_xdr(:base64)
+
+      expect { 
+        sep10.read_challenge_tx(challenge_xdr: challenge, server: server)          
+      }.to raise_error(
+        Stellar::InvalidSep10ChallengeError,
+        /The transaction has expired/
+      )
+
+      envelope.tx.time_bounds = Stellar::TimeBounds.new(min_time: 0, max_time: 5)
+      challenge = envelope.tx.to_envelope(server, user).to_xdr(:base64)
+
+      expect { 
+        sep10.read_challenge_tx(challenge_xdr: challenge, server: server)          
+      }.to raise_error(
+        Stellar::InvalidSep10ChallengeError,
+        /The transaction has expired/
+      )
+
+      now = Time.now.to_i
+      envelope.tx.time_bounds = Stellar::TimeBounds.new(
+        min_time: now + 100, 
+        max_time: now + 500
+      )
+      challenge = envelope.tx.to_envelope(server, user).to_xdr(:base64)
+
+      expect { 
+        sep10.read_challenge_tx(challenge_xdr: challenge, server: server)          
+      }.to raise_error(
+        Stellar::InvalidSep10ChallengeError,
+        /The transaction has expired/
+      )
+    end
+  end
+
+  describe "#verify_challenge_tx_threshold" do
```

Comment:
All of these tests test functionality defined in other functions. 

`verify_challenge_tx_threshold` is essentially a call to `verify_challenge_tx_signers` with the additional functionality of checking `threshold`. I guess this is more a question of testing philosophy, but I don't see the value in adding these tests for this function.

Even if `verify_challenge_tx_threshold` no longer called `verify_challenge_tx_signers` down the line, whatever it did call instead would presumably also have tests for its functionality. If we moved the implementation of said functionality into `verify_challenge_tx_threshold`, thats when it would be appropriate to add the unit tests there. In my opinion tests should be applied to wherever the functionality is implemented.

Reply:
I often take the approach you describe, but the reason we were so thorough with tests in the other SDKs is because these functions are handling signature verification and authentication, so there is a lot at stake. The tests I listed are tests we have in the other SDKs.

Reply:
I missed asking for two tests on the `_threshold` function, which it looks like we also missed in the Go SDK, and strangely enough would catch a double counting of weight bug that I think exists in this function:

- [ ] duplicate signers
- [ ] duplicate signatures